### DISCUSSION TITLE
how to unlock intercom from function or express endpoint

### DISCUSSION THREAD

**Discussion by gabr981** [dgreif/ring#1529] [Q&A]
### I can't unlock the device, it's not clear which ID to use, I passed the two IDs I could find but nothing doesn't work...

```

app.post('/open-gate/:intercomID', async (req, res) => {

    try {
        const locations = await ringApi.getLocations();
        const intercoms = locations.flatMap(location => location.intercoms || []);
        const intercom = intercoms.find(i => i.data.device_id == req.params.intercomID);
        
        try {
            await intercom.unlock();
        } catch (error) {
            return res.status(500).json({ error: 'Errore interno del server', details: error });
        }

        res.json({ status: 'Porta sbloccata con successo', deviceName: intercoms[1].id, deviceType: intercom.name });
 
 
        if (!intercom) {
            return res.status(404).json({ error: 'Dispositivo intercom non trovato' });
        }

        if (typeof intercom.unlock !== 'function') {
            return res.status(400).json({ error: 'Il dispositivo non supporta la funzione unlock' });
        }

    } catch (error) {
        console.error('Errore in /open-gate:', error);
        res.status(500).json({ error: 'Errore interno del server', details: error.message });
    }

});
```

**Comment by tsightler** [comment]
Not much context here so a bit difficult to understand the request.  Overall, I don't see anything obviously wrong in the code above.  The ID passed in the request should be whatever matches the i.data.device_id for that device so that the correct device can be located, then unlock() is called on that device.

**Reply by gabr981** [reply]
when I run try catch with unlock function, it goes to catch with thi error


```
"response": {
      "headers": {},
      "status": 404,
      "body": {
        "error": "Client Device with e08341d6-85fb-5b53-a3bd-128b13e793ae not found"
      }
    }
  }
}
```

## I also edited the code
```
app.post('/open-gate/:intercomID', async (req, res) => {

    try {
        const locations = await ringApi.getLocations();
        //const client = new RingIntercom();
        const intercoms = locations.flatMap(location => location.intercoms || []);
        const intercom = intercoms.find(i => i.id == req.params.intercomID);
        //const intercom = intercoms[0];
        const rclinet = new RingRestClient({
            refreshToken: process.env.RING_REFRESH_TOKEN,
            cameraStatusPollingSeconds: 20,
            locationModePollingSeconds: 20,
        });
        const ringIntercom = new RingIntercom(intercom, rclinet);
        //return res.status(200).json({ details: intercom.id, locations: locations[0].intercoms[0].id, restClient: rclinet});
        

        try {
            let intercom2 = intercom.device_id;
            await intercom.unlock();
        } catch (error) {
            return res.status(500).json({ error: 'Errore interno del server', details: error });
        }

        res.json({ status: 'Porta sbloccata con successo', deviceName: intercoms[1].id, deviceType: intercom.name });
 
 
      

    } catch (error) {
        console.error('Errore in /open-gate:', error);
        res.status(500).json({ error: 'Errore interno del server', details: error.message });
    }

```

**Reply by tsightler** [reply]
The code you just posted is different from the code you posted earlier, which only adds to the confusion.  I have no idea why you are trying to create a new rest client when you already have a client, that makes no sense.  And you are using a different ID, so even more confusing.  None of what you've added is required and you've made it way more complex for some reason.

**Reply by gabr981** [reply]
## I was doing some tests here the clean code

```
app.post('/open-gate/:intercomID', async (req, res) => {

    try {
        const locations = await ringApi.getLocations();
        const intercoms = locations.flatMap(location => location.intercoms || []);
        const intercom = intercoms.find(i => i.id == req.params.intercomID);
        
        try {
            await intercom.unlock();
            res.json({ status: 'Porta sbloccata con successo', deviceName: intercoms[1].id, deviceType: intercom.name });
        } catch (error) {
            return res.status(500).json({ error: 'Errore interno del server', details: error });
        }

    } catch (error) {
        console.error('Errore in /open-gate:', error);
        res.status(500).json({ error: 'Errore interno del server', details: error.message });
    }

});

```

#### This code gives the same error and the ID I pass is different than the error one

**Reply by tsightler** [reply]
I don't understand how the code you write can produce the response you state.  You say it is dropping to a catch block, but both catch blocks return only status 500 so how are the response codes being returned?  There have been occasional reports of 404 with Intercom unlock in the past (last one was almost a year ago), but it usually goes away in a day or two.  Can you provide the full output of the intercom device object?  Maybe just add `console.error(JSON.stringify(intercom))` after the intercom is found?

**Reply by gabr981** [reply]
### I print the error in the "details" key of the catch and add a new key to print the intercom object. The `error` is under the details key.


### Case 1
```
        try {
            //let intercom2 = intercom.device_id;
            await intercom.unlock();
        } catch (error) {
            return res.status(500).json({ error: 'Errore interno del server', details: error, intercom: JSON.stringify(intercom) });
        }
        
        
        
        
        
        
        {
  "error": "Errore interno del server",
  "details": "Converting circular structure to JSON\n    --> starting at object with constructor 'Timeout'\n    |     property '_idlePrev' -> object with constructor 'TimersList'\n    --- property '_idleNext' closes the circle"
}
```

### Case 2

```

        try {
            //let intercom2 = intercom.device_id;
            await intercom.unlock();
        } catch (error) {
            return res.status(500).json({ error: 'Errore interno del server', details: error });
        }



{
  "error": "Errore interno del server",
  "details": {
    "response": {
      "headers": {},
      "status": 404,
      "body": {
        "error": "Client Device with e08341d6-85fb-5b53-a3bd-128b13e793ae not found"
      }
    }
  }
}
```

**Reply by tsightler** [reply]
Yes, I didn't expect you to return the intercom object in the error message, I just wanted to output enough of it to be sure it is finding the correct device, so we could see the intercom id and device_id (they are different).  The ID format in the message doesn't match what I would expect to see from an intercom.  They are usually much shorter and not formatted like a UUID.

Also, are you using the primary Ring account or some secondary account?  Obviously the account must have the ability to perform an unlock, I'm not sure if shared accounts allow this.

**Reply by gabr981** [reply]
Yes, I understand your concern. When I print intercom.data, the ID appears to be correct and matches what I expect. However, the issue arises when I attempt to use the unlock() method. I’ve tried filtering both by id and device_id, but neither approach resolves the issue.

Regarding the account, I am using the primary Ring account, the same account used for the initial configuration, so permissions should not be an issue.

**Reply by tsightler** [reply]
How you filter doesn't really matter, as long as it is the correct device that you are calling unlock on, it should work.  None of us as developers own an intercom, so we can't really test, but the code here is identical to code in python-ring-doorbell and there are quite a significant number of intercom users of ring-homebridge, ring-mqtt and Ring integration in HA, which all use this method, so it's unlikely to be wrong as the various forums would be full of reports, so I'm not really sure what else to tell you to try.  You could try adding some manual debugging to the unlock function in ring-client-api and see what is happening there, is it generating the correct URL, etc.  But, overall, this is very basic stuff, not really sure what it could be.

**Reply by gabr981** [reply]
Thank you for your feedback. I actually already use this library with Home Assistant, and it works perfectly there. However, I’m working on a self-check-in project where I need to exclude Home Assistant, as I don’t require all its functionalities.

I don’t think the issue lies with the library it’s more likely that I’m making a mistake somewhere in my implementation. However, I haven’t been able to pinpoint the error yet. I’ll continue debugging and let you know if I find anything unusual. Thank you for your suggestions!
